Refactor

mysterycommand opened the pull request:
## This is a Nearly Complete Refactor
Though the interface and functionality is the same, the module has been refactored to make testing easier, and adding a 'require hook' possible.

### Major changes:
* Moved the main library files into `./lib/`.
  * `./index.js` just exports `./lib/sassr.js`.
* `./lib/sassr.js` has been rewritten to work more like [`jstify`](https://github.com/zertosh/jstify), it exposes the main transform by default, but also:
  * ~`compile` - does some `opts` cleanup and normalization, but otherwise passes through to [`node-sass.render`](https://github.com/sass/node-sass#render-callback--v300)~
  * `transform` - does the main work of what old-Sassr did, taking the result of `compile` and wrapping it in a Node module (either an error module or a working CSS module).
    * **NOTE:** this used to be an inline string concatenation, but now just uses a Lodash template for each type of module.
  * `sync` - a synchronous version of the main export (I dunno, might come in handy one day).
  * ~`compileSync` - a synchronous version of compile (passes opts to [`node-sass.renderSync`](https://github.com/sass/node-sass#render-callback--v300)~
  * `transformSync` - a synchronous version of transform.
* Added a 'require hook' available via `require('sassr/register')();` … that call can also take options that are passed through to the transform (with the exception of `data` which is generated "on the fly" for whatever Sass file you're trying to load … obv).
  * **NOTE:** this 'require hook' makes use of `transformSync` and `compileSync`. The `sync` version of the Browserify transform is mainly there for internal consistency/completeness and isn't currently in use.
* There are tests of the main lib functionality. Not 100% sure how/if we want to test the 'require hook', but there's also an `./examples` folder that demonstrates that it does indeed work.

**NOTE:** I just removed the compile functions, the were mainly just passthroughs anyway.

kmck commented on lib/sassr.js:
```diff
@@ -0,0 +1,164 @@
+}
+
+module.exports = sassr;
+module.exports.compile = compile;
```
i'd probably add the functions onto `sassr` directly instead of `module.exports`

mysterycommand commented:
@kmck like this?:
```
sassr.sync = sassrSync;
sassr.transform = transform;
sassr.transformSync = transformSync;

module.exports = sassr;
```

shz commented:
Doing a clean checkout of this branch and an `npm install` gives me this error when I run `npm test`:

```
{ [Error: module "sassr/lib/style-element-helper" not found from "/Users/lpstein/Development/Yahoo/sassr/test/mock/style.css"]
  filename: 'sassr/lib/style-element-helper',
  parent: '/Users/lpstein/Development/Yahoo/sassr/test/mock/style.css' }
```

shz commented:
And, documenting this in one place:

These two in combination don't do a complete job of escaping the output, which can lead to an error if there's e.g. `'` in a stylesheet:
https://github.com/johnnybenson/sassr/pull/7/files#diff-4948d6232bc0809a98f355d55b2cc104R34
https://github.com/johnnybenson/sassr/pull/7/files#diff-dbd7f3f0015e7d991e49a3ddfaf523fdR4

For example, using this:

```css
.blah { font-family: 'Helvetica'; }
```

Results in this JS result:

```js
'use strict';

var style = require('sassr/lib/style-element-helper');
var css = '.blah{font-family:'Helvetica'}\n'; // Yikes!
var appended;

exports.getStyleElement = function() {
    return style.element;
};

exports.getCSSText = function() {
    return css;
};

exports.append = function() {
    if (!appended) {
        style.inject(css);
        appended = true;
    }

    return style.element;
};

exports.remove = function() {
    if (appended) {
        style.eject(css);
        appended = false;
    }

    return style.element;
};

```

mysterycommand commented:
Re: the initial test failure, it's because the generated CSS modules require that helper, but they're already running from *inside* `sassr`. The only thing I could think of was to `npm link && npm link sassr` into the project (just to get the tests to run) … weird, but what else to do? Require itself!?

Re: the single quote thing … yeah, that makes sense now, the original code used double quotes and I was like "ew" … but `JSON.stringify` should escape nested doubles, and wrap it in doubles. Double quotes for the CSS I guess ("ew" though).

shz commented:
On quotes: looks like the original used the JSON trick because there was no other great solution.  But since `lodash` is doing templating now, can it be escaped using that rather than "rolling your own" like the original code?

shz commented:
On the test thing... rather than doing the `npm link`s, can it be accomplished by checking in a symlink?

mysterycommand commented:
Okay, I don't think there's a Lodash escaping helper that will work for this and it looks like double-quotes and `JSON.stringify` works well enough.

Also, after talking it over with @shz it seems like a symlink won't really work because it'd have to link `./` into `./node_modules` (unless there's some other way to pick `sassr` in `$NODE_PATH`) and that seems dicey to me.

kmck commented:
after playing around with this, i'm thinking that browserify isn't the best way to test this. i have some decent tests here: https://github.com/johnnybenson/sassr/compare/kmck/require-hook

here are my thoughts:
* ugly as they are, i'm cool with keeping the concatenated strings right now for the output template. i'd rather keep this simple and fast
* let's not introduce jstify in here
* `jsdom` is no longer supported in node, so that's a dealbreaker for that dependency.
* generally i don't think there's a ton of value in trying to mimic the directory structure from `npm init`, but i don't mind `lib/` for the meat of the library
* not super into the directory structure for the tests the way you have it here. i changed it on my branch and it's pretty clean.

anyway, the commit history on this branch is getting pretty hairy, so i think maybe keep tweaking this until you get it to the state you want (for the library itself, don't waste a bunch of time on the tests), i wanna start a fresh branch with a commit history that we'll be able to look at and understand.

kmck commented:
weirdly, i think maybe i didn't have your latest changes went i went through this branch, so i still wanna take another look. if you could squash those commits, it'd reaaaaaaally make it easier to follow the changes being made :grin:

mysterycommand commented:
Okay @kmck, hopefully this is easier to parse. In response to your earlier notes:

1. after playing around with this, i'm thinking that browserify isn't the best way to test this. i have some decent tests here: https://github.com/johnnybenson/sassr/compare/kmck/require-hook
  * Cool, just use a stream, test the stream, I like it.
2. ugly as they are, i'm cool with keeping the concatenated strings right now for the output template. i'd rather keep this simple and fast
  * Okay, I'll do that.
3. let's not introduce `jstify` in here
  * Sure, it's only actually used here in the `./examples` folder to verify that the latest version (with the require hook) works with such an old version of Browserify (it seems to).
4. `jsdom` is no longer supported in node, so that's a dealbreaker for that dependency.
  * Eh, `jsdom@3` works fine in Node, and Node 4 will effectively *be* Io.js. Anyway, "dealbreaker" seems overly strong, but it looks like the `vm`/context stuff you have in your branch works.
5. generally i don't think there's a ton of value in trying to mimic the directory structure from `npm init`, but i don't mind lib/ for the meat of the library
  * Does `npm init` give you a directory structure? I thought it just gave you a `package.json`.
6. not super into the directory structure for the tests the way you have it here. i changed it on my branch and it's pretty clean.
  * You just renamed mocks to fixtures, got rid of the Browserify-related stuff, and moved the tests "up" a level … ¯\_(ツ)_/¯.

johnnybenson commented:
hey at some point `append` and `remove` methods were renamed to `inject` and `eject` you gotta update the README

mysterycommand commented:
@johnnybenson the interface that's exposes by the generated module still uses `append` and `remove`, but the internal helper functions use `inject` and `eject`.
